### DISCUSSION TITLE
Ability to show workflow state/steps and current progress

### DISCUSSION THREAD

**Discussion by bryanvam** [elsa-workflows/elsa-core#6012] [General]
I have currently created a fixed step workflow. Defined programmatically. I am using the IWorkflowExecutionLogStore to get the workflow history in order to populate the below. I am wanting a more dynamic approach. Can anyone point me in the right direction? This is especially relevant in the case where a workflow can go back to the begining or a previous step. I need to show the outcome of the step - for example a comment from a manager why the workflow cannot proceed. They need to see that comment in the relevant step.

![image](https://github.com/user-attachments/assets/1c1e40f6-ecac-4ca1-8f81-2567bb3faef4)

**Comment by aubryll** [comment]
@bryanvam Did you find a solution to this?

**Reply by bryanvam** [reply]
not out of the box. i had to build it from scratch

**Reply by aubryll** [reply]
Could you share any pointers?

**Comment by sfmskywalker** [comment]
Short answer: there still isn't a built-in "current progress" projection for rendering a stepper like the screenshot, but the runtime APIs now give you better primitives than relying on `IWorkflowExecutionLogStore` alone.

If you want to build this dynamically, I would look at these first:
- `IWorkflowInstancesApi.GetJournalAsync` / `GetFilteredJournalAsync` for the instance journal.
- `IWorkflowInstancesApi.GetExecutionStateAsync` for overall workflow status/sub-status.
- `IWorkflowInstancesApi.GetVariablesAsync` for workflow-level state that you want to surface in the UI.
- `IActivityExecutionsApi.ListAsync` / `ListSummariesAsync` / `GetReportAsync` for per-activity execution state, bookmark presence, timing, and summaries.

A useful detail here is that `WorkflowExecutionLogRecord.ActivityState` is now marked obsolete in the runtime model, with the code pointing you to `ActivityExecutionRecord.ActivityState` instead. So if you need a per-step projection, activity execution records are the better long-term source.

Relevant source:
- `elsa-core/src/clients/Elsa.Api.Client/Resources/WorkflowInstances/Contracts/IWorkflowInstancesApi.cs`
- `elsa-core/src/clients/Elsa.Api.Client/Resources/ActivityExecutions/Contracts/IActivityExecutionsApi.cs`
- `elsa-core/src/modules/Elsa.Workflows.Runtime/Entities/WorkflowExecutionLogRecord.cs`
- `elsa-core/src/clients/Elsa.Api.Client/Resources/ActivityExecutions/Models/ActivityExecutionRecord.cs`
- `elsa-core/src/modules/Elsa.Workflows.Api/Endpoints/ActivityExecutions/Report/Endpoint.cs`

For comments like "manager rejected this step because ...", I would not depend on the journal alone. I'd store that data explicitly in workflow variables or your own domain/persistence model, then correlate it back to `ActivityId` / `ActivityNodeId` in your projection.

For waiting/blocking steps, `HasBookmarks` on activity execution records is also useful when you want to show that a step is paused pending input.

**Comment by sfmskywalker** [comment]
Closing this as answered. There still is not a built-in stepper/progress projection for this UI, but the current `IWorkflowInstancesApi` and `IActivityExecutionsApi` endpoints are the right primitives for building one from journal, execution state, variables, and per-activity execution records.
